src/cipher/pelz_aes_gcm.c

```diff
@@ -0,0 +1,395 @@
+  {
+    free(*cipher);
+    free(*iv);
+    free(*tag);
+    *cipher = NULL;
+    *iv = NULL;
+    *tag = NULL;
+    *cipher_len = 0;
+    *iv_len = 0;
+    *tag_len = 0;
+    EVP_CIPHER_CTX_free(ctx);
+    return 1;
+  }
+
+  // now that the encryption is complete, clean-up cipher context
+  EVP_CIPHER_CTX_free(ctx);
+
+  return 0;
+}
+
+//############################################################################
+// aes_gcm_decrypt()
+//############################################################################
+int pelz_aes_gcm_decrypt(unsigned char *key,
+			 size_t key_len,
+			 unsigned char* iv,
+			 size_t iv_len,
+			 unsigned char* cipher,
+			 size_t cipher_len,
+			 unsigned char* tag,
+			 size_t tag_len,
+			 unsigned char** plain,
+			 size_t* plain_len)
+{
+  // validate non-NULL and non-empty decryption key specified
+  if (key == NULL || key_len == 0)
+  {
+    return 1;
+  }
+
+  // validate non-NULL and non-empty input ciphertext buffer specified
+  if (cipher == NULL || cipher_len == 0)
+  {
+    return 1;
+  }
+
+  if(iv == NULL || iv_len == 0)
+  {
+    return 1;
+  }
+
+  if(tag == NULL || tag_len == 0)
+  {
+    return 1;
+  }
+  
+  *plain_len = cipher_len;
+  *plain = malloc(*plain_len);
+  if(*plain == NULL)
+  {
+    *plain_len = 0;
+    return 1;
+  }
+  
+  // variables to hold/accumulate length returned by EVP library calls
+  //   - OpenSSL insists this be an int
+  int len = 0;
+  int plaintext_len = 0;
+
+  // initialize the cipher context to match cipher suite being used
+  EVP_CIPHER_CTX *ctx;
+
+  if (!(ctx = EVP_CIPHER_CTX_new()))
+  {
+    free(*plain);
+    *plain = NULL;
+    *plain_len = 0;
+    return 1;
+  }
+  int init_result = 0;
+
+  switch (key_len)
+  {
+  case 16:
+    init_result = EVP_DecryptInit_ex(ctx, EVP_aes_128_gcm(), NULL, NULL, NULL);
+    break;
+  case 24:
+    init_result = EVP_DecryptInit_ex(ctx, EVP_aes_192_gcm(), NULL, NULL, NULL);
+    break;
+  case 32:
+    init_result = EVP_DecryptInit_ex(ctx, EVP_aes_256_gcm(), NULL, NULL, NULL);
+    break;
+  default:
+    break;
+  }
+  if (!init_result)
+  {
+    free(*plain);
+    *plain = NULL;
+    *plain_len = 0;
+    EVP_CIPHER_CTX_free(ctx);
+    return 1;
+  }
+
+  // set tag to expected tag passed in with input data
+  if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_TAG, tag_len, tag))
+  {
+    free(*plain);
+    *plain = NULL;
+    *plain_len = 0;
+    EVP_CIPHER_CTX_free(ctx);
+    return 1;
+  }
+
+  // set the IV length in the cipher context
+  if (!EVP_CIPHER_CTX_ctrl(ctx, EVP_CTRL_GCM_SET_IVLEN, iv_len, NULL))
+  {
+    free(*plain);
+    *plain = NULL;
+    *plain_len = 0;
+    EVP_CIPHER_CTX_free(ctx);
+    return 1;
+  }
+
+  // set the key and IV in the cipher context
+  if (!EVP_DecryptInit_ex(ctx, NULL, NULL, key, iv))
+  {
+    free(*plain);
+    *plain = NULL;
+    *plain_len = 0;
+    EVP_CIPHER_CTX_free(ctx);
+    return 1;
+  }
+
+  // decrypt the input ciphertext, put result in the output plaintext buffer
```

Comment:
Similar to the encrypter, there are 5 opportunities for this function to return after plain is allocated and before use.

Reply:
I moved the allocation to right before it's needed in 3b51da7.

Reply:
In 5829857 I moved setting default values for the output variables to the top of the functions.